src/lib/kademlia/membership.ml

```diff
@@ -54,6 +54,43 @@ let get_project_root () =
   in
   go @@ realpath current_dir_name
 
+(* This snippet was taken from our fork of RPC Parallel.
+ * Would be nice to have a shared utility, but this is
+ * easiest for now. *)
+(* To get the currently running executable:
+  On Darwin:
+  Use _NSGetExecutablePath via Ctypes
+
+  On Linux:
+  Use /proc/PID/exe
+   - argv[0] might have been deleted (this is quite common with jenga)
+   - `cp /proc/PID/exe dst` works as expected while `cp /proc/self/exe dst` does not *)
+let get_coda_binary =
+  let our_binary_lazy =
+    lazy
+      (let open Async in
+      let open Deferred.Or_error.Let_syntax in
+      let%map os = Process.run ~prog:"uname" ~args:["-s"] () in
+      if os = "Darwin\n" then (
+        let open Ctypes in
+        let ns_get_executable_path =
+          Foreign.foreign "_NSGetExecutablePath"
+            (ptr char @-> ptr uint32_t @-> returning void)
+        in
+        let path_max = 1024 in
+        let buf = Ctypes.allocate_n char ~count:path_max in
+        let count =
+          Ctypes.allocate uint32_t (Unsigned.UInt32.of_int (path_max - 1))
+        in
+        ns_get_executable_path buf count ;
+        let s =
+          string_from_ptr buf ~length:(!@count |> Unsigned.UInt32.to_int)
+        in
+        List.hd_exn @@ String.split s ~on:(Char.of_int 0 |> Option.value_exn) )
```

Comment:
I can't believe Ctypes doesn't have helpers for dealing with C strings, but apparently it doesn't. :(

Reply:
Yes, unfortunate, I know.